Fix issue 151

Karim-Mane opened the pull request:
This PR contains changes to adress issue #151

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -66,11 +77,22 @@ scan_columns <- function(x) {
+  # to prevent from the effects of the conversion to numeric and character.
+  #
+  types         <- as.character(vapply(data, class, character(1L)))
+  scan_result   <- NULL
+  j             <- 1
+  for (i in names(data)) {
+    scan_result <- rbind(scan_result, scan_columns(data[[i]], types[j]))
+    j           <- j + 1
+  }
```
data.frames, matrices & vectors should never be grown in R.

In other words, `for` loops should never rely on iteratively using `c()` or `rbind()`.

```suggestion
  scan_result <- vapply(seq_len(ncol(data)), function(col_index) {
    scan_columns(data[[col_index]], types[[col_index]])
   }, numeric(5))
```

Karim-Mane replied:
I have account for this in commit `828534d`.

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -18,8 +18,12 @@ scan_columns <- function(x) {
-  tmp         <- suppressWarnings(as.numeric(x))
-  are_numeric <- round((sum(!is.na(tmp)) / n_rows), 6L)
+  are_numeric <- 0L
+  verdict     <- type == "logical" | type == "factor"
```
```suggestion
  verdict     <- type == "logical" || type == "factor"
```

Karim-Mane replied:
I accounted for this. Thanks

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -66,11 +77,22 @@ scan_columns <- function(x) {
+  # The percent of numeric and character value will be set automatically to 0 as
+  # to prevent from the effects of the conversion to numeric and character.
+  #
+  types         <- as.character(vapply(data, class, character(1L)))
```
Do we want to use `class()` or `typeof()` here?

```suggestion
  types         <- vapply(data, class, character(1L))
```

Karim-Mane replied:
`class()` will gives us the inofrmation that we need - the composition in data types of every column.

`typeof()` will tell us date values are double, numeric. This is not what we want to spot out from the `scan_data()` function.

Bisaloo replied:
But `class()` doesn't always return an output of length 1 so this will error.

You will see this with a test including `POSIXct` columns for example

Karim-Mane replied:
I still believe that `typeof()` is not appropriate for this task.

I have made changes to account for multiple types in commit `db420c8`.

Bisaloo reviewed (changes requested):
I have added some technical comments but I still don't understand the logic.

Could you add a comment at the start of `scan_columns()` to explain please?

E.g.,:

- what about numeric values encoded as factors?
- can a value be of type logical but in reality be a date?
- are you sure dates won't be counted twice?

``` r
as.numeric(as.Date("2020-01-01"))
#> [1] 18262
```

<sup>Created on 2024-07-23 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Please add more examples & tests with more diverse column types.

Bisaloo commented:
Please add more examples & tests with more diverse column types.

In particular, there is still an issue with dates.

Karim-Mane commented:
> Please add more examples & tests with more diverse column types.
> 
> In particular, there is still an issue with dates.

I have tested the function across multiple datasets. If you have a dataset with odd cases that are not accounted for yet, please share this with me.

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -1,54 +1,145 @@
+#' @return A character with the class of the input vector
+#' @keywords internal
+#'
+get_class <- function(x) {
+  paste(class(x), collapse = ",")
+}
```
In general, it's a bit idea to manipulate objects / code as text, especially to circumvent explicit safeguards

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -1,54 +1,145 @@
-  x        <- x[which(is.na(tmp))]
+  # get the proportion of date values
   are_date <- 0L
+  if (any(type %in% c("POSIXct", "POSIXt"))) {
```
This solution doesn't scale as we can't possibly add special cases for every existing types. Especially since packages can define their own types.

Bisaloo reviewed (changes requested):
I don't think this is going in the right direction.

The code is getting more and more complicated, which will inevitably lead to more bugs.

I would recommend:

- centering the different case treatment on `typeof()` which has a finite, known, set of outputs
- think about what possible `scan_data()` types are possible for each `typeof()` and create custom `scan()` function for each type. For example:
   - it seems that logical can only lead to logical and NA, why is it grouped with factors?
   - character are on the other extreme of the spectrum since something that appears as a character can actually be any of the types recognized by `scan_data()`

It may be easier to start from a blank page to not be influenced by the current code, which seems to push toward more and more complexity.

Karim-Mane commented:
> I don't think this is going in the right direction.
> 
> The code is getting more and more complicated, which will inevitably lead to more bugs.
> 
> I would recommend:
> 
>     * centering the different case treatment on `typeof()` which has a finite, known, set of outputs
> 
>     * think about what possible `scan_data()` types are possible for each `typeof()` and create custom `scan()` function for each type. For example:
>       
>       * it seems that logical can only lead to logical and NA, why is it grouped with factors?
I had identified two special cases (logical abd factors) that I wanted to resolve using their level. But can be splitted into two small functions.
>       * character are on the other extreme of the spectrum since something that appears as a character can actually be any of the types recognized by `scan_data()`.
Character columns can contain data of any types. If you start by determining the proportion of characters in this function, most columns will appear to only contain characters. Detecting the other types first, then assigning the remaining to characters was the best alternative I could think of. Happy to explore any other suggestion around this.
> 
> 
> It may be easier to start from a blank page to not be influenced by the current code, which seems to push toward more and more complexity. Pen & paper design, with a flow diagram of the different pathways, may help.

I will detach myself from this for a day or two. I will give it a try with `typeof()` when I get back to it after.

Bisaloo commented:
Maybe as a first question to determine the best way forward: what is the exact role of this function in a data cleaning pipeline? How is its output used to inform following steps? Why these specific types have been selected as potential types recognized by `scan_data()`?

Karim-Mane commented:
> Maybe as a first question to determine the best way forward: what is the exact role of this function in a data cleaning pipeline? How is its output used to inform following steps? Why these specific types have been selected as potential types recognized by `scan_data()`?

> what is the exact role of this function in a data cleaning pipeline?
If by `pipeline` here you meant a suit of cleaning operations that are executed one after the other by passing the output of one as an input for the next function, the `scan_data()` function will not be used within such pipeline.
         It is intended to be used prior the application of any cleaning operations. The aim here is to provide the user with a comprehensive content of every column in the dataset in terms of the proportion of the defined data types. 

> How is its output used to inform following steps? 
For example, a column with both numeric and character values might require a call for `convert_to_numeric()`, or a conversion of the numeric values into character, or setting the values of one of these type to `NA`. With this information in hand, the user can plan the set of cleaning operations to apply on the dataset.

> Why these specific types have been selected as potential types recognized by `scan_data()`?
The selected types are: `numeric, date, character, logical`. This is purely based on observation - they are the common types found in most of dataset I have encountered.
Note that here `date` is a combination of both `Date` and `date-time` (POSITct and POSITlt).

> How to adjust this function?
These atomic vector types are built on top of the 4 basic types (`interger, double, logical, character`). We will define a function that makes use of the `typeof()` function. Based on the returned value of this function, We will determine the percent of `factor, Date, date-time, character, numeric, and logical` in every column of the dataset.

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -65,12 +38,185 @@ scan_columns <- function(x) {
+    scan_columns(data[[col_index]], types[[col_index]])
+  }, numeric(7L))
+  scan_result        <- as.data.frame(t(scan_result))
+  names(scan_result) <- c("missing", "numeric", "date", "character",
+                          "logical", "date-time", "factor")
```
Why did this change? It doesn't give the impression that the list results from an explicit and motivated design choice?

Bisaloo replied:
It seems that this function is now mostly returning the `class()`. Can we go back to the drawing board and write a list of requirements:

- when I have X type of data, I want to know Y to inform Z

> For example, a column with both numeric and character values might require a call for convert_to_numeric(), or a conversion of the numeric values into character, or setting the values of one of these type to NA. With this information in hand, the user can plan the set of cleaning operations to apply on the dataset.

This is a good example for analyzing types in `character` columns. What about the others?

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -65,12 +38,185 @@ scan_columns <- function(x) {
+
+  # doubles are either numeric (attributes = NULL), or Date (has a 'class'
+  # attributes = Date), or date-time (has a 'class' attributes = POSIXt)
+  if ("class" %in% names(attributes(x))) {
+    if ("Date" %in% attributes(x)[["class"]]) {
+      are_date <- round((length(x) / n_rows), 6L)
+    } else if ("POSIXt" %in% attributes(x)[["class"]]) {
+      are_date_time <- round((length(x) / n_rows), 6L)
+    }
```
As mentioned, this is not the proper way to test class in R. Use `inherits()`. Please don't use complex workarounds to go around this.

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -65,12 +38,185 @@ scan_columns <- function(x) {
+#' @keywords internal
+#'
+scan_in_double <- function(x) {
+  are_factor <- are_date <- are_date_time <- are_character <- are_numeric <-
```
Please do not initialize values when not necessary

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -65,12 +38,185 @@ scan_columns <- function(x) {
+      are_date_time <- round((length(x) / n_rows), 6L)
+    }
+  } else {
+    are_numeric <- round((length(x) / n_rows), 6L)
```
So 1? Can we write this instead?

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -65,12 +38,185 @@ scan_columns <- function(x) {
+#'    that were detected within the input vector.
+#' @keywords internal
+#'
+scan_in_double <- function(x) {
```
In a nutshell, this function is running `class()` and counting NA. Do we need to make it as complex?

Bisaloo reviewed (changes requested):
The most important criteria in all the code we produce is that it should be clear and maintainable.

When you write code, please go through it to double check if that is the case.

In general, this will happen by using standard mechanisms (e.g., to test class) and by being explicit about the values of object. E.g., if the value is 1, let's write one.

If standard patterns or mechanisms are unknown to you, a good strategy is to start following a GitHub repo of a known package and look at the changes they are making.

Karim-Mane commented:
@Bisaloo - Thanks for taking the time to review this. 

I disagree with some comments above.
We have been working on this and other packages. When there is some code that is unclear or non-maintainable, please be specific as you used to do up to now. When the comments are this vague and written in a certain way, it is demoralising.

 Let's meet tomorrow to clarify our conceptions of the function and remove all ambuguities around it.

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
+    date_count <- date_count + sum(!is.na(tmp))
+
+    # convert to numeric and check for the presence of Date among the numeric
+    tmp2 <- x[is.na(tmp)]
```
Is it possible to use more expressive names than `tmp`?

Karim-Mane replied:
I have accounted for this in commit `9c8f69f`.

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
+  date_count <- numeric_count <- 0L
+  if (sum(!is.na(tmp)) > 0L) {
+    # Setting the first date to 50 years before the current date
+    target_interval <- sort(
+      seq.Date(Sys.Date(), length.out = 2L, by = "-50 years")
+    )
```
Only the first value is used thereafter. What about creating a simpler `oldest_possible_date` object which would be a `Date` of length 1?

Karim-Mane commented:
> Thanks! I think it's going in a good direction. I have left specific technical comments but here are also some general reminders to address at some point before this gets merged:
> 
>     * [x]  Add an explanation in design vignettes and/or function documentation about in which scenarios this function could be useful
> 
>     * [x]  Add an explanation about why only character columns are analyzed
> 
>     * [ ]  Add an explanation about the rationale behind output format (normalized to 1 or not? Row- vs col-orientation? etc.)

Thanks for quoting these. 
I have added some text to the design vignette and updated the function documentation. 
I could not however provide a rationale behind the use of proportion vs percentage. I guess I am used to proportions...

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,222 +1,147 @@
+#' @returns A data frame with the same number of rows as the number of character
+#'    columns of the input data, and six (06) columns representing the field
+#'    names, the percentage of missing, numeric, date, character, and logical,
+#'    values in each column.
```
This is out of habits. I have removed `06` from the this text. Will check for such occurence in other files of the package and harmonise.

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
+  # send an message if there is no character column found within the input data
+  if (length(target_columns) == 0L) {
+    message("No character column found in the provided data.")
+    return(invisible(NA))
```
The message would have been enough for this. I am returning `NA` here so that in order to have a value that will be used to know whether the scanning result will be displayed or not in the `print_report()` function.
`Invisibly` because I am masking the NA from the user's console as the message is informative enough.

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
+  scan_result <- vapply(seq_len(ncol(data)), function(col_index) {
+    scan_in_character(data[[col_index]])
+  }, numeric(5L))
+  scan_result        <- as.data.frame(t(scan_result))
```
The reasoning behind this format is primarly due to the display in the report generated by `print_report()`.
For a dataset with many character columns, the user will have to scroll going to the right-side if the output of `scan_data()` is displayed as you shown above.
With cols -> rows, the first fifthteen columns are shown and the user can click on the available buttons to see the rest. This is simpler to my opinion - but I am happy to adjust if there is strong arguments for the other format.

Bisaloo replied:
Thanks, rotating to avoid scrolling seems like a good reason! Can you make it explicit in the design vignette please?

Something like:

> We transpose the result relative to the input dataset (columns in the input are returned as row) to avoid horizontal scrolling in the case of datasets with a large number of character columns.

Bisaloo commented:
>  I could not however provide a rationale behind the use of proportion vs percentage. I guess I am used to proportions...

To clarify, I am not talking about proportion vs percentage.

I'm saying that we could potentially not sum to 1 (or 100%). Given the use case you described, maybe we don't care if we double count some values in different categories (e.g. `date` and `numeric`).

We should maybe(?) return a simple "can it be converted to...?" without trying to rely on complex heuristics to differentiate between date & numeric. With this, the user can decide if the entire column should be date or numeric.

Does this make more sense?

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
+
+  # scan through the character columns
+  data <- data[, names(target_columns)]
+  scan_result <- vapply(seq_len(ncol(data)), function(col_index) {
+    scan_in_character(data[[col_index]])
+  }, numeric(5L))
```
I have accounted for this in commit `81da4e7`.

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
+  }
+
+  # scan through the character columns
+  data <- data[, names(target_columns)]
```
I had noticed this from the example that involves a linelist.

I have accounted for this in commit `81da4e7`.

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
-                                   are_date + are_logical)), 6L)
+  # get the character count
+  character_count <- initial_length -
+    (na_count + logical_count + numeric_count + date_count)
 
-  # --- return the output ---
-  return(c(are_na, are_numeric, are_date, are_character, are_logical))
-}
+  # transform into proportions
+  props <- round(
+    c(na_count, numeric_count, date_count, character_count, logical_count) /
+      initial_length, 4L)
```
The guessing part only happens when `Date` values are identified from the parsing using `lubridate::parse_date_time()`. Otherwise, all numeric values are strictly considered as numeric. I think, the following could help users:

1. Add a note about this in the function description
2. Add the algorithm used within the function in the `details` section of the function documentation.

Bisaloo replied:
With this in mind, I believe my question is still valid: should we do **any** guessing at all? If we don't have a strong reason to, I'd suggest we don't.

For example, if we have "20210107", should it be a date or a numeric? For a Covid-19 dataset, it could be either. It's probably safer to leave the final choice to the user.

Karim-Mane replied:
Here is another proposition from @Degoot-AM 

I would add a data type called `ambiguous`  and assigned to it non-differentiable entries and that would add the sum of raw to 1 (100%).

Karim-Mane commented:
> > I could not however provide a rationale behind the use of proportion vs percentage. I guess I am used to proportions...
> 
> To clarify, I am not talking about proportion vs percentage.
> 
> I'm saying that we could potentially not sum to 1 (or 100%). Given the use case you described, maybe we don't care if we double count some values in different categories (e.g. `date` and `numeric`).
> 
> We should maybe(?) return a simple "can it be converted to...?" without trying to rely on complex heuristics to differentiate between date & numeric. With this, the user can decide if the entire column should be date or numeric.
> 
> Does this make more sense?

This is related to the opened conversation above. Can we follow up this topic in that conversation?

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
+    return(invisible(NA))
+  }
+
+  # unclass the data to prevent from warnings when dealing with linelist, and
+  # scan through the character columns
```
This type of comment is exactly the kind of thing we need :+1:. Thanks!

Karim-Mane replied:
I would have reply to this in wolof if you knew the language...

Bisaloo commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,147 @@
+
+  # unclass the data to prevent from warnings when dealing with linelist, and
+  # scan through the character columns
+  data               <- as.data.frame(unclass(data))[, target_columns]
```
Dropping the linelist class can more easily be achieved by.

```suggestion
  data               <- as.data.frame(data)[, target_columns]
```

Bisaloo reviewed (commented):
I left two minor comments inline.

I think we're almost there:

- let's make sure all the points discussed throughout this process are in the design vignette
- let's finalize the conversation of double counting

Karim-Mane commented:
> I left two minor comments inline.
> 
> I think we're almost there:
> 
>     * let's make sure all the points discussed throughout this process are in the design vignette
> 
>     * let's finalize the conversation of double counting

I have accounted for the documentation and double counting in commit `5b405f9`.

Bisaloo reviewed (changes requested):
I don't think this is behaving as intended:

```r
scan_in_character("20210702", "test")
```

should return `c(0, 1, 1, 0, 0)`

```r
scan_in_character(c("20210702", "2021/07/03", "3"), "test")
```

should return `c(0, 0.6666, 0.6666, 0, 0)`

It would probably be helpful to add tests that test exact results, in a stricter fashion that just testing class or length.

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,179 @@
+  } else {
+    are_numeric <- suppressWarnings(as.numeric(x))
+    numeric_count <- numeric_count + sum(!is.na(are_numeric))
+    character_count <- character_count - numeric_count
```
I have accounted for this in commit `1cf3d3c`.

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,179 @@
+  date_count <- numeric_count <- 0L
+  if (sum(!is.na(are_date)) > 0L) {
+    # Setting the first date to 50 years before the current date
+    oldest_date <- seq.Date(Sys.Date(), length.out = 2L, by = "-50 years")[[2L]]
```
accounted for this in commit `fa1d9d7`.

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,179 @@
-  are_character <- round((1.0 - (are_na + are_numeric +
-                                   are_date + are_logical)), 6L)
+    # get the date count
+    date_count <- date_count + sum(!is.na(are_date))
```
I have accounted for this

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,179 @@
+    are_numeric <- suppressWarnings(as.numeric(non_date))
+    character_count <- character_count - sum(!is.na(are_numeric))
+    numeric_count <- numeric_count + sum(!is.na(are_numeric))
+    if (sum(!is.na(are_numeric)) > 0L) {
```
I have accounted for this

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,179 @@
+    }
+  } else {
+    are_numeric <- suppressWarnings(as.numeric(x))
+    numeric_count <- numeric_count + sum(!is.na(are_numeric))
```
I have accounted for this

Bisaloo commented on tests/testthat/test-clean_data_helpers.R:
```diff
@@ -1,11 +1,52 @@
+  # 2 date values (one of them `20210702` is also numeric)
+  # 2 numeric values in which one is also a date value, hence the
+  # warning about the presence of ambiguous data
+  dat <- data.frame(col1 = c(c("20210702", "2021/07/03", "3"), "test"))
```
```suggestion
  dat <- data.frame(col1 = c("20210702", "2021/07/03", "3", "test"))
```

Bisaloo reviewed (commented):
I can't post it as a suggestion but I believe this much simpler implementation should work:

```r
scan_in_character <- function(x, x_name) {
  # There might be, within a character column, values of type:
  # character, numeric, date (date or date-time), NA, and logical
  # In this function, we check the presence of these different types within a
  # character column.

  # save the variable length
  # the character count is decreased by the number of occurrence a different
  # data type is found.
  initial_length <- character_count <- length(x)

  # get the count of missing data (NA)
  na_count <- sum(is.na(x))
  x      <- x[!is.na(x)]
  character_count <- character_count - na_count

  # We will check if there is any Date values within the variable by parsing the
  # values, looking for the ones that fit any of the predefined format.
  #     When there is one or more Date values, they will be converted into
  # numeric. The first numeric count is recorded at this point. The rest of the
  # values are converted into numeric, and the second count of numeric is
  # recorded. They will in turn be converted into date.
  # If any of these numeric values is a Date (a numeric, which
  # after conversion to Date, fall within the interval
  # [50 years back from today's date, today's date]), it will add to the second
  # date count.
  # That way the Date count is the count of date identified from the
  # parsing + the count of Dates within the numeric values. Similarly, the
  # numeric count is the count of numeric values within dates values and count
  # among the non-date values.
  #
  # NOTE: This is what justifies that the sum across rows in the output object
  # could be > 1.
  #
  #     When there is no Date values identified from the parsing, the variable
  # is converted into numeric. The final numeric count is the sum of all the
  # identified numeric values.
  #     The logical count is the number of TRUE and FALSE written in both lower
  # and upper cases within the variable.
  #     The remaining values will be considered of type character.

  # parsing the vector, looking for date values
  are_date <- suppressWarnings(
    as.Date(
      lubridate::parse_date_time(
        x,
        orders = c("ymd", "ydm", "dmy", "mdy", "myd", "dym", "Ymd", "Ydm",
                   "dmY", "mdY", "mYd", "dYm")
      )
    )
  )

  date_count <- sum(!is.na(are_date))
  character_count <- character_count - date_count

  # convert everything to numeric and get the numeric count
  are_numeric <- suppressWarnings(as.numeric(x))
  numeric_count <- sum(!is.na(are_numeric))

  numeric_and_date_count <- sum(!is.na(are_date) & !is.na(are_numeric))

  if (numeric_and_date_count > 0) {
    cli::cli_alert_warning(
      "Found {numeric_and_date_count} value{?s} that can be both numeric and \\
      date in column `{x_name}`"
    )
  }

  # We don't want to remove numeric_and_date values twice
  character_count <- character_count - numeric_count + numeric_and_date_count

  # get logical count
  logicals <- toupper(x) == "TRUE" | toupper(x) == "FALSE"
  logical_count <- sum(logicals)
  character_count <- character_count - logical_count

  # transform into proportions
  props <- round(
    c(na_count, numeric_count, date_count, character_count, logical_count) /
      initial_length, 4L
  )

  return(props)
}
```

Please let me know if you prefer that I push it to this branch or as a PR so you can review it more easily.

Karim-Mane commented on R/clean_data_helpers.R:
```diff
@@ -1,76 +1,216 @@
+
+    # send a warning about the number of ambiguous values found on that column
+    if (ambiguous_count > 0) {
+      cli::cli_alert_warning(
+        "Found {ambiguous_count} values that can be either numeric or date in",
+        "column `{x_name}`"
+      )
```
@Bisaloo - Please create a PR.

Karim-Mane commented:
Closing this PR. The requested changes have been implemented on [fix-issue-151-simplification](https://github.com/epiverse-trace/cleanepi/tree/fix-issue-151-simplification)`, which is now merged with main.
